higher-order-functions/writing-higher-order-functions.md

```diff
@@ -0,0 +1,171 @@
+
+Now, let's demystify the `min` function with the optional `key` parameter by writing our own custom min function `min_function_custom`. We will use `get_first_movie` as a reference:
+
+```py
+def get_first_movie(movies):
+    min_movie = movies[0]
+    min_title = movies[0]["title"]
+    for movie in movies:
+        if movie["title"] < min_title:
+            min_movie = movie
+            min_title = movie["title"]
+    return min_movie
+```
+
+Notice that we can replace each line with the expression `movie["title"]` with a function call `get_title(movies[0])` without changing the behavior of the function. 
+
+```py
+def get_title(movie):
+    return movie["title"]
+
+
+def get_first_movie(movies):
+    min_movie = movies[0]
+    min_title = get_title(movies[0])
+    for movie in movies:
+        if get_title(movie) < min_title:
+            min_movie = movie
+            min_title = get_title(movie)
+    return min_movie
+```
+
+Now, let's take it one step further to generalize our function to find the minimum movie by any key (i.e. `"title"`, `"rating"`, or `"release_year"`). 
+
+We will add a second parameter, a function that gets the value for a specified key: `get_value_for_key`. We will generalize our variable names changing `movies` to `collection` and `movie` to `item.` Each line from `get_first_movie` that has the expression `get_title(movie)` will be replaced with `get_value_for_key(item)`.
+
+
+| Specific Function | Generalized Function |
+| -- | -- |
+| `movies` | `collection`|
+| `movie` | `item` |
+| `get_title(movie)` | `get_value_for_key(item)` |
+| `def get_first_movie(movies)`|`def min_function_custom(collection, get_value_for_key)`|
+*Table. Relationship between variables in specific function an generalized function.*
+
+```py
+def min_function_custom(collection, get_value_for_key):
+     min_item = collection[0]
+     min_value = get_value_for_key(collection[0])
+     for item in collection:
+         if get_value_for_key(item) < min_value:
+             min_item = item
+             min_value = get_value_for_key(item)
+     return min_item
+```
+
+We can now class `min_function_custom` just like the built in `min` function. We can pass in a named function or a lambda expression for the `get_value_for_key` parameter
+
+```py
+first_movie_by_title = min_function_custom(movies, get_value_for_key=get_title)
+lowest_rated_movie = min_function_custom(movies, get_value_for_key=get_rating)
+earliest_movie = min_function_custom(
+    movies, get_value_for_key=get_release_year)
+```
+or 
+
+```py
+first_movie_by_title = min_function_custom(
+    movies, get_value_for_key=lambda movie: movie["title"])
+lowest_rated_movie = min_function_custom(
+    movies, get_value_for_key=lambda movie: movie["rating"])
+earliest_movie = min_function_custom(
+    movies, get_value_for_key=lambda movie: movie["release_year"])
+```
+
+We can also pass the second parameter as a positional argument rather than a keyword argument:
```

Comment:
I might have missed it, but I don't recall seeing an example of moving the order of arguments with keyword arguments here or in the previous lesson that introduced them. I think it'd be nice to have a callout around "keyword arguments can be passed in a different order from how they're defined, but that cannot be done with positional"